aiida/repository/backend/abstract.py

```diff
@@ -84,15 +84,33 @@ def put_object_from_file(self, filepath: typing.Union[str, pathlib.Path]) -> str
             return self.put_object_from_filelike(handle)
 
     @abc.abstractmethod
+    def has_objects(self, keys: List[str]) -> List[bool]:
+        """Return whether the repository has an object with the given key.
+
+        :param keys:
+            list of fully qualified identifiers for objects within the repository.
+        :return:
+            list of logicals, in the same order as the keys provided, with value True if the respective
+            object exists and False otherwise.
+        """
+
     def has_object(self, key: str) -> bool:
         """Return whether the repository has an object with the given key.
 
         :param key: fully qualified identifier for the object within the repository.
         :return: True if the object exists, False otherwise.
         """
+        return self.has_objects([key])[0]
+
+    @abc.abstractmethod
+    def list_objects(self) -> Iterable[str]:
+        """Return iterable that yeilds all available objects by key.
+
+        :return: An iterable for all the available object keys.
+        """
 
     @contextlib.contextmanager
-    def open(self, key: str) -> typing.Iterator[typing.BinaryIO]:
+    def open(self, key: str) -> Iterator[BinaryIO]:
```

Comment:
```suggestion
    def open(self, key: str) -> ContextManager[BinaryIO]:
```

A bit more specific: https://docs.python.org/3/library/typing.html#typing.ContextManager (I only noticed this was available myself recently)

Reply:
actually scratch that, I think this is only for without the contextmanager decorator (at least mypy/pylint are raising a lot of errors